Code review

vincent1114 opened the pull request:
Using pull request to code review

MTAN2802 committed:
Made grid, fixed layout

MTAN2802 committed:
Added buttons, fixed formatting of grid

MTAN2802 committed:
Added scaling for different screen dimensions

MTAN2802 committed:
Create functions for start + reset buttons

MTAN2802 committed:
Fixed format of text when starting game

MTAN2802 committed:
Create object for tiles

MTAN2802 committed:
Fixed isThereAGame function

MTAN2802 committed:
Made function to shift first row to right

MTAN2802 committed:
Made "D" move first row right.

MTAN2802 committed:
Create function to show open squares

MTAN2802 committed:
Fixed code to add a new "2" tile after each turn

MTAN2802 committed:
Added path where 2 merges with another 2

MTAN2802 committed:
Added comment to each if statement for moveRight

MTAN2802 committed:
Completed moveRight function for "2" tiles

MTAN2802 committed:
Fixed move function; 3rd and 4th if statement
conditions.
E.g. for last row 3rd if condition ends in d < 15
and 4th ends in d < 16.
Added <p> element to numbers that were missing it

MTAN2802 committed:
Fixed issue where tile would not go right as
intended

MTAN2802 committed:
Added if statement so moveRight function only
works if there is a tile on the board

MTAN2802 committed:
Update 2048-game.js

Created function for moving right, left, up, and down

MTAN2802 committed:
Fixed error where "2" tiles was merging with "32"

MTAN2802 committed:
Fixed issue where number was being added when
tiles were being shifted to a side with no space

MTAN2802 committed:
a

MTAN2802 committed:
Added function to determine whether tiles should
be able to move or not

MTAN2802 committed:
Changed canItMoveRight to only work if there is
a way to move right

MTAN2802 committed:
Fixed all functions that determine if tiles
can move in each direction

MTAN2802 committed:
Update 2048-game.js

Fixed function that moves tiles to right
- Goes from right to left now when combining tiles

MTAN2802 committed:
Merge pull request #1 from MTAN2802/MTAN2802-patch-1

Update 2048-game.js

MTAN2802 committed:
Fixing another issue with shifting tiles;
Need to change it so the function goes through each
tile instead of each number

MTAN2802 committed:
Added function to determine which number was in the selected tile based on innerHTML;
Changed shiftRight function to not use parameters, goes through each tile from right to left, bottom to top;
Used new function to get current tile's number and colour and add it by 1 to get the next number

MTAN2802 committed:
Fixed all shifting functions

MTAN2802 committed:
Renamed to index.html

MTAN2802 committed:
Create README.md

MTAN2802 committed:
Create _config.yml

vincent1114 commented on 2048-game.css:
```diff
@@ -0,0 +1,170 @@
+    font-size: 48px;
+}
+
+@media only screen and (max-width: 1440px){
```
damn my guy doing responsive already

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+const fourteen = document.getElementById("fourteen");
+const fifteen = document.getElementById("fifteen");
+const sixteen = document.getElementById("sixteen");
+const arr = [one, two, three, four, five, six, seven, eight, nine, ten, eleven, twelve, thirteen, fourteen, fifteen, sixteen];
```
Try to use a more meaningful variable names.
Using number as variable names is hard to understand what this really mean.
Better name would be tile1, tile2 etc.

Which you could've used numbers instead of characters instead of the div ids too

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+     colour: "currantRed",
+     hex: "#680C07",
+     numCode: "<p>2048</p>"}
+]
```
Look up what prettier is, it formats your code for you.
Also this is probably better with a hashmap/object-map instead of a list

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+     numCode: "<p>2048</p>"}
+]
+
+//Functions for the buttons
```
If you think functions belong in a collection, you can make a new file and import them so the code is easy to read and edit

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+            }
+        }
+    }
+}
```
Always better to return early than use a if else statement
```
function initiateGame(){
    let test = 14;
    let randomNo = Math.floor(Math.random() * 16);
    if (isThereAGame()){
        alert('There is already an active game!')
        return;
    }
    for (i=0; i < arr.length; i++){
        if (test == i){
            arr[i].style.backgroundColor = `${tiles[0].hex}`;
            arr[i].innerHTML = `<p>${tiles[0].number}</p>`;
        }
    }
}
```

But I don't get what this function is doing
what is test
why use a loop for the else statement when you're just trying to assign one thing
isn't `arr[test].style.backgroundColor ....` good enough

```
function initiateGame(){
    let test = 14;
    let randomNo = Math.floor(Math.random() * 16);
    if (isThereAGame()){
        alert('There is already an active game!')
        return;
    }
    arr[test].style.backgroundColor = `${tiles[0].hex}`;
    arr[test].innerHTML = `<p>${tiles[0].number}</p>`;
}
```

MTAN2802 replied:
I was testing how the tiles would move at specific tiles, forgot to get rid of the test variable

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+function canItMoveRight(){
+    let counter = 0;
+    for (i = arr.length-1; i >=0; i--){
+        if (arr[i].style.backgroundColor != ""){
```
Unless optimising, it's a bad idea to use css properties or any properties that does not directly represent the meaning. You can use a better data structure for the game array. (Also a better name)

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+    let counter = 0;
+    for (i = arr.length-1; i >=0; i--){
+        if (arr[i].style.backgroundColor != ""){
+            if (i == 3 || i == 7 || i == 11 || i == 15){
```
You can do 
`if ([3, 7, 11, 15].includes(i)) { `

in python it would be `if i in [3,7,11,15]: ` which makes more human sense

Then you have a variable called edges the next line, you can use that variable for this so it becomes
`if (edges.include(i))` which is a lot cleaner and self documents what it's doing

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+            if (i == 3 || i == 7 || i == 11 || i == 15){
+                let edges = [3,7,11,15];
+                edges.forEach((edgeNo) => {
+                    if (arr[edgeNo - 3].style.backgroundColor != "" && arr[edgeNo - 3].innerHTML == arr[edgeNo - 2].innerHTML){
```
This is very hard to understand what we're doing
Practice commenting complex logic.

Using else statement for this entire block probably makes more sense too, if each edge only meant to plus one counter.

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+                })
+                    
+            }
+            if(i != 3 && i != 7 && i != 11 && i != 15){
```
this looks like it can be an else statement

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+        }
+    }
+    //console.log(`Right counter: ${counter}`)
+    if (counter != 0) return true;
```
you can just `return count !== 0`

You can also return early by replacing any of the `counter++` to `return true`
and `return false;` in the end. This is more efficient so you don't have to go through the rest of loop.

ALSO, always use brackets for multi-line if statements. Some people even don't like inline if statements but I find it alright

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+    else return false;
+}
+
+//Function to see which number is in selected tile
```
This comment is not necessary, your function name is good here so it self documents.

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+    return n
+}
+
+//Function to move right
```
Same here, but this shiftRight name is a bit vague, so use a longer name or have a better comment, i prefer having a better/longer function name

vincent1114 replied:
Should explain the logic of the function on top of the pretty good comments you have on each if statement

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+                const secondRightColumn = [2, 6, 10, 14];
+                const x = numInTile(arr[i].innerHTML);
+
+                if (secondRightColumn.includes(i)){ //If the selected tile is in the second right column
```
damn you are using include
good

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+
+                if (secondRightColumn.includes(i)){ //If the selected tile is in the second right column
+                    if (arr[i+1].style.backgroundColor == ""){ //If the very right column is empty
+                        arr[i+1].style.backgroundColor = arr[i].style.backgroundColor;
```
These blocks can probably be a function

vincent1114 commented on 2048-game.js:
```diff
@@ -0,0 +1,777 @@
+function moveUp(event){
+    if (isThereAGame()){
+        if (canItMoveUp()){
+            if (event.key == "w" || event.key == "W"){
```
can easily add the arrow keys

vincent1114 commented on 2048-game.js:
It seems like a lot of your logic is repeated, try to encapsulate and use more functions.

```
if (tileExists(i) && tileEmpty(i + 4)) {
    moveTile(i, i + 4);
}
```

Something like this makes everything more readible, and if you design code this way it usually is easier

You can separate functions out in different files that do a collection of things.

vincent1114 reviewed (commented):
Just a rough review, I didn't look through the logic too closely
Most of your functions have a lot of the same problems so I only made comment on the first few

MTAN2802 committed:
Changed div id to be numbers, changed variable name for tiles.
Added arrow key event listeners

MTAN2802 committed:
Added "add2TileOr4Tile" function, used return early instead of if else

add2TileOr4Tile function is for determining whether the new number added will be 2 or 4.
70% chance of 2; 30% chance of 4

MTAN2802 committed:
Made new function to see if tiles can move in a direction

MTAN2802 committed:
sdadada

MTAN2802 committed:
Fixed functions to see if tiles could move for all directions

MTAN2802 committed:
Changed tile details to object maps

Added currentValue attribute to prevent using css attributes in shifting functions later on

MTAN2802 committed:
Created function isEmpty

MTAN2802 committed:
Fix functions

Not finished

MTAN2802 committed:
Shortened code

MTAN2802 committed:
Fixed new functions, I think it works now

MTAN2802 committed:
Merge pull request #3 from MTAN2802/MTAN2802-patch-2

Fix functions

MTAN2802 committed:
Old stuff

MTAN2802 committed:
Brought back stuff
